core/domain/draft_upgrade_services_test.py

```diff
@@ -314,25 +406,41 @@ def test_convert_states_v28_dict_to_v29_dict(self):
                 'property_name': 'content',
                 'new_value': 'new value'
             })]
-        self.assertEqual(
-            draft_upgrade_services.DraftUpgradeUtil._convert_states_v28_dict_to_v29_dict(  # pylint: disable=protected-access,line-too-long
-                draft_change_list)[0].to_dict(),
-            draft_change_list[0].to_dict())
+        # Migrate exploration to state schema version 29.
+        self.perform_and_assert_exploration_states_schema_migration('28', '29')
+        # Migrate the draft change list's state schema to the migrated
+        # exploration's schema.
+        migrated_draft_change_list_v29 = (
+            draft_upgrade_services.try_upgrading_draft_to_exp_version(
+                draft_change_list, 1, 2, self.EXP_ID)
+        )
+        self.assertEqual(migrated_draft_change_list_v29[0].to_dict(),
+                         draft_change_list[0].to_dict())
 
     def test_convert_states_v27_dict_to_v28_dict(self):
-        draft_change_list = [
+        draft_change_list_v27 = [
             exp_domain.ExplorationChange({
                 'cmd': exp_domain.CMD_EDIT_STATE_PROPERTY,
                 'property_name': 'content_ids_to_audio_translations',
                 'state_name': 'State B',
                 'new_value': 'new value',
             })]
-        self.assertEqual(
-            draft_upgrade_services.DraftUpgradeUtil._convert_states_v27_dict_to_v28_dict(  # pylint: disable=protected-access,line-too-long
-                draft_change_list)[0].to_dict(),
+        # Version 28 adds voiceovers_mapping.
+        draft_change_list_v28 = [
             exp_domain.ExplorationChange({
                 'cmd': exp_domain.CMD_EDIT_STATE_PROPERTY,
                 'property_name': 'recorded_voiceovers',
                 'state_name': 'State B',
                 'new_value': {'voiceovers_mapping': 'new value'}
-            }).to_dict())
+            })]
+        # Migrate exploration to state schema version 28.
+        self.perform_and_assert_exploration_states_schema_migration('27', '28')
+        # Migrate the draft change list's state schema to the migrated
+        # exploration's schema.
+        migrated_draft_change_list_v28 = (
+            draft_upgrade_services.try_upgrading_draft_to_exp_version(
+                draft_change_list_v27, 1, 2, self.EXP_ID)
+        )
+        self.assertEqual(
+            migrated_draft_change_list_v28[0].to_dict(),
```

Comment:
Here and elsewhere, why are you comparing just [0]? You should be checking exact equality -- perhaps check `[change.to_dict() for change in ...]` with each other.

In general, if you are referring to specific indexes in what should be fairly generic code, that's probably a sign that you need to think about how to write the code a bit more generally.

Reply:
Yes. That's because your current check isn't verifying that the new change list contains the same number of elements as the old one -- so, the approach I suggested is more generic and comprehensive.